opencxl/cxl/device/cxl_type1_device.py

```diff
@@ -201,11 +219,92 @@ def _init_device(
+            if isinstance(pckt, CxlCacheH2DDataPacket):
+                # received a data packet
+                # since we're ignoring the 32B transfer case,
+                # we can just assume this data packet contains everything we want
+                fut_data.set_result(pckt.data)
+                raise CancelledError
+            if (
+                pckt.h2drsp_header.cache_opcode != CXL_CACHE_H2DRSP_OPCODE.GO
+                or pckt.h2drsp_header.rsp_data
+                in (CXL_CACHE_H2DRSP_CACHE_STATE.INVALID, CXL_CACHE_H2DRSP_CACHE_STATE.ERROR)
+            ):
+                raise CancelledError  # terminate the listener and free the cqid
+
+        await self._cxl_cache_manager.send_d2h_req_rdown(addr)
+
+        if not cqid:
+            cqid = await self.get_next_cqid()
+
+        # avoid sequential cacheline writes by maintaining callbacks which are
+        # executed upon retrieval of a Rsp with matching cqid.
+
+        self._cxl_cache_manager.register_cqid_listener(
+            cqid=cqid,
+            cb=_listen_check_set_fut,
+            _timeout=20,
+        )
+
+        return fut_data
+
+    async def cxl_cache_writeline(self, addr: int, data: int, cqid: Optional[int] = None):
+        def _listen_check_send(pckt: CxlCacheH2DRspPacket):
+            """
+            Callback registered to listen for the given CQID.
+            Checks if the host response packet represents an error.
+            If not, sends the requested cacheline write.
+            """
+            # don't need to perform the type check twice
+            # just trust in ourselves!
+            if pckt.h2drsp_header.cache_opcode == CXL_CACHE_H2DRSP_OPCODE.GO_ERR_WRITE_PUL:
+                raise CancelledError  # terminate the listener and free the cqid
+            self._cxl_cache_manager.send_d2h_data(data, pckt.h2drsp_header.rsp_data)
+
+        await self._cxl_cache_manager.send_d2h_req_itomwr(addr, cqid)
+
+        if not cqid:
+            cqid = await self.get_next_cqid()
+
+        # avoid sequential cacheline writes by maintaining callbacks which are
+        # executed upon retrieval of a Rsp with matching cqid.
+        self._cxl_cache_manager.register_cqid_listener(
+            cqid=cqid,
+            cb=_listen_check_send,
+            _timeout=20,
+        )
+
+    async def cxl_cache_readlines(self, addr: int, howmuch: int) -> int:
+        # pylint: disable=not-an-iterable
+        CACHELINE_LENGTH = 64
+        lines = bytearray(max(howmuch, 64))
+        async for l_idx in range(math.ceil(howmuch / CACHELINE_LENGTH)):
+            lines[l_idx * CACHELINE_LENGTH : (l_idx + 1) * CACHELINE_LENGTH] = bytes(
+                await self.cxl_cache_readline(
+                    addr + l_idx * CACHELINE_LENGTH, await self.get_next_cqid()
+                )
+            )
+        return int.from_bytes(lines)
 
     async def cxl_cache_writelines(self, addr: int, data: int, howmuch: int):
```

Comment:
Same here. Though, this seems to be old code. Please modify as well.

Reply:
Could you clarify what you mean by old code? This is a new function.

Reply:
As in, this line is not part of this commit since it used to be line 207 (not line 301) and:
```
async def cxl_cache_writelines(self, addr: int, data: int, howmuch: int):
    pass
```
Usually only the "changes" are discussed in code reviews, but since this is very closely related to the new code, I was asking for it to be changed as well.